http-clients/src/main/java/com/palantir/remoting3/clients/UserAgents.java

```diff
@@ -88,83 +87,76 @@ private static String formatSingleAgent(UserAgent.Agent agent, Map<String, Strin
-        if (primaryAgent.isPresent()) {
-            builder.primary(primaryAgent.get());
-        } else {
+        if (!foundFirst) {
             if (lenient) {
-                log.warn("Malformed primary user agent", UnsafeArg.of("agent", parts.get(0)));
-                builder.primary(UserAgent.Agent.of("unknown", UserAgent.Agent.DEFAULT_VERSION));
+                log.warn("Empty user agent, falling back to default/unknown agent");
+                return builder.primary(UserAgent.Agent.of("unknown", UserAgent.Agent.DEFAULT_VERSION)).build();
             } else {
-                throw new IllegalArgumentException("Agent string was malformed: " + parts.get(0));
+                throw new IllegalArgumentException("Failed to parse user agent string: " + userAgent);
             }
         }
-        if (primaryComments.containsKey("nodeId")) {
-            builder.nodeId(primaryComments.get("nodeId"));
-        }
 
-        // Informational agents
-        for (String part : parts.subList(1, parts.size())) {
-            Optional<UserAgent.Agent> informationlAgent = tryParseSingleAgent(part, Maps.newHashMap());
-            if (informationlAgent.isPresent()) {
-                builder.addInformational(informationlAgent.get());
-            } else {
-                if (lenient) {
-                    // Ignore malformed agent
-                    log.warn("Cannot parse malformed user agent string", UnsafeArg.of("agent", part));
-                } else {
-                    throw new IllegalArgumentException("Agent string was malformed: " + part);
-                }
-            }
-        }
         return builder.build();
     }
 
-    // returns found comments in the provided map.
-    private static Optional<UserAgent.Agent> tryParseSingleAgent(String agent, Map<String, String> comments) {
-        int slashPos = agent.indexOf('/');
-        if (slashPos == -1) {
-            return Optional.empty(); // no version
-        }
-        int spacePos = agent.indexOf(" (");
-        String name = agent.substring(0, slashPos);
-        String version = agent.substring(slashPos + 1, spacePos == -1 ? agent.length() : spacePos);
-
-        // Extract comments
-        if (spacePos != -1) {
-            String commentString = agent.substring(spacePos + 2, agent.length() - 1);
-            COLON_SEMICOLON_SPLITTER.split(commentString).forEach(comments::put);
+    private static Map<String, String> parseComments(String commentsString) {
+        Map<String, String> comments = Maps.newHashMap();
+        for (String comment : COMMENT_SPLITTER.split(commentsString)) {
+            List<String> fields = COLON_SPLITTER.splitToList(comment);
+            if (fields.isEmpty()) {
+                // continue
+            } else if (fields.size() == 2) {
+                comments.put(fields.get(0), fields.get(1));
+            } else {
+                comments.put(comment, comment);
```

Comment:
MapSplitter doesn't work because of this case?

Reply:
yep